### DISCUSSION TITLE
How to speed up PyVista rendering when adding many meshes?

### DISCUSSION THREAD

**Discussion by leszeknowakowski** [pyvista/pyvista#6404] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
import sys
import numpy as np
import pyvista as pv
from PyQt5 import QtWidgets, QtCore
from pyvistaqt import QtInteractor


class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()

        self.setWindowTitle("PyVista Sphere Plotter")
        self.setGeometry(100, 100, 800, 600)

        self.central_widget = QtWidgets.QWidget()
        self.setCentralWidget(self.central_widget)

        self.layout = QtWidgets.QVBoxLayout()
        self.central_widget.setLayout(self.layout)

        self.slider = QtWidgets.QSlider(QtCore.Qt.Horizontal)
        self.slider.setMinimum(1)
        self.slider.setMaximum(100)
        self.slider.setValue(60)  # initial radius * 100 (for 0.6)
        self.slider.setTickInterval(1)
        self.slider.setTickPosition(QtWidgets.QSlider.TicksBelow)
        self.layout.addWidget(self.slider)

        self.slider.valueChanged.connect(self.update_radius)

        # PyVista
        self.plotter = QtInteractor(self.central_widget)
        self.layout.addWidget(self.plotter.interactor)

        self.plot_spheres(0.6)

    def plot_spheres(self, radius):
        self.plotter.clear()
        centers = np.random.rand(110, 3) * 10  # Random centers for the spheres

        for center in centers:
            sphere = pv.Sphere(radius=radius, center=center, theta_resolution=10, phi_resolution=10)
            self.plotter.add_mesh(sphere, color='blue')

        self.plotter.reset_camera()

    def update_radius(self, value):
        radius = value / 100.0
        self.plot_spheres(radius)


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    window = MainWindow()
    window.show()
    #sys.exit(app.exec_())
```

### Description

Hello, 
I like how PyVista wraps around vtk, but in my case it comes at the cost of performance. Is there any way to speed up the rendering? Consider attached code - rendering a 110 random spheres and changing their radius with a slider. With PyVista it becomes sluggerish, being slower even 4 - 5 times than VTK, heavily affecting the interaction. Profiling those 2 examples (vtk_test is attached below) reveals that PyVista calls method "Render" of "vtkmodules.vtkRenderingOpenGLRenderWindow 112 times, but pure VTK code only 1 time. How to avoid this obstacle within PyVista ( I really don't want switching back to VTK), but still using add_mesh method (or any other method that will add an actor to a renderer - lot of my code depends on these actors) ?

#1678  is not the solution since it adds only one actor..

[vtk_test.txt](https://github.com/user-attachments/files/16326197/vtk_test.txt)

Thanks in advance

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Sun Jul 21 23:54:13 2024 Środkowoeuropejski czas letni

                OS : Windows
            CPU(s) : 12
           Machine : AMD64
      Architecture : 64bit
               RAM : 31.1 GiB
       Environment : Python
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : NVIDIA GeForce RTX 4070/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 551.23
  MathText Support : False

  Python 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64
  bit (AMD64)]

           pyvista : 0.43.3
               vtk : 9.3.0
             numpy : 1.26.4
        matplotlib : 3.8.2
            scooby : 0.9.2
             pooch : 1.8.0
            pillow : 10.2.0
         pyvistaqt : 0.11.0
             PyQt5 : 5.15.10
           IPython : 8.20.0
             scipy : 1.12.0
        jupyterlab : 4.0.11
      nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

**Comment by user27182** [comment]
> ```python
>     def plot_spheres(self, radius):
>         self.plotter.clear()
>         centers = np.random.rand(110, 3) * 10  # Random centers for the spheres
> 
>         for center in centers:
>             sphere = pv.Sphere(radius=radius, center=center, theta_resolution=10, phi_resolution=10)
>             self.plotter.add_mesh(sphere, color='blue')
> 
>         self.plotter.reset_camera()
> 
>     def update_radius(self, value):
>         radius = value / 100.0
>         self.plot_spheres(radius)
> ```

This code completely redraws the plot every time `update_radius` is called. The geometry for all the spheres is generated every time and re-mapped for rendering which is expensive. Instead, you should create the plot with all the spheres just once, and only _update_ the spheres without creating new ones. To perform the update you have many options:

1. You can use `pv.SphereSource` instead of `pv.Sphere` and in `update_radius()` you would modify the `radius` of each source directly and call `Update()` to update the spheres
2. You can continue using `pv.Sphere`, but in `update_radius()` you would modify the datasets directly using a transformation, e.g. `dataset.scale()`
3. Similar to 2,  you continue using `pv.Sphere`, but only use a single sphere and instead create many actors. In `update_radius()`, you re-scale the actors directly.

For example, using option 3 might look something like:

``` python
# Only create a single sphere and set initial center and radius
sphere_dataset = pv.Sphere(radius=1, center=(0,0,0), theta_resolution=10, phi_resolution=10)

# Create many actors which use the same sphere
n_spheres = 100
centers = np.random.rand(n_spheres, 3) * 10  # Random centers for the spheres
actors = []
for center in centers:
    # Create new actor
    actor = pv.Actor(mapper=pv.DataSetMapper(dataset=sphere_dataset))
    # Position actor at desired location
    actor.position = center
    
    # Keep track of actors
    actors.append(actor)

    # Initialize plot with actors 
    self.plotter.add_actor(actor, color='blue')


def update_radius(self, value):
    # Update radius by re-scaling the actor
    scale = value / 100.0
    for actor in actors:
        actor.scale = scale
```

Note that I haven't tested the code above, this is just an idea/suggestion. The main thing is to first _initialize_ the plot, then only update the actors instead of completely re-making the plot from scratch.

**Reply by leszeknowakowski** [reply]
Great, thanks for answer! I managed to get to work method 3, and it is really much faster. The problem is solved, but aparat from that, this approach still calls vtkmodules.vtkRenderingOpenGLRenderWindow.Render() each time we call plotter.addActor() method from pyvista. This is seiously slowing the initializng the plot, and - as we can see from vtk example - it is sufficient to call Render() only once, after adding all of the actors. The solution  is add kwarg "render=False" to add_mesh or add_actor method and then append plotter.update(). Now it adds all actors, but render them only once at the end. Thank You for help!
